src/core/solution.jl

```diff
@@ -119,8 +128,250 @@ function add_pipe_resistance_setpoint(sol, wm::GenericWaterModel)
     end
 end
 
+""
+function add_setpoint_node_head!(sol, wm::AbstractWaterModel)
+    add_setpoint!(sol, wm, "nodes", "h", :h,
+        status_name=wm_component_status["nodes"],
+        inactive_status_value=wm_component_status_inactive["nodes"])
+end
+
+function add_setpoint_node_head!(sol, wm::AbstractCNLPModel)
+    add_dual!(sol, wm, "nodes", "h", :flow_conservation,
+        scale=(x, item) -> -x, status_name=wm_component_status["nodes"])
+end
+
+""
+function add_setpoint_pipe_flow!(sol, wm::AbstractWaterModel)
+    add_setpoint!(sol, wm, "pipes", "q", :q,
+        status_name=wm_component_status["pipes"],
+        inactive_status_value=wm_component_status_inactive["pipes"])
+end
+
+""
+function add_setpoint_pump_flow!(sol, wm::AbstractWaterModel)
+    add_setpoint!(sol, wm, "pumps", "q", :q,
+        status_name=wm_component_status["pumps"],
+        inactive_status_value=wm_component_status_inactive["pumps"])
+end
+
+function add_setpoint_pipe_resistance!(sol, wm::AbstractWaterModel)
+    if InfrastructureModels.ismultinetwork(wm.data)
+        data_dict = wm.data["nw"]["$(wm.cnw)"]["pipes"]
+    else
+        data_dict = wm.data["pipes"]
+    end
+
+    sol_dict = get(sol, "pipes", Dict{String, Any}())
+
+    if length(data_dict) > 0
+        sol["pipes"] = sol_dict
+    end
+
+    if :x_res in keys(var(wm))
+        for (i, link) in data_dict
+            a = link["index"]
+            sol_item = sol_dict[i] = get(sol_dict, i, Dict{String, Any}())
+
+            if a in keys(var(wm, :x_res))
+                x_res, r_id = findmax(JuMP.value.(var(wm, :x_res, a)))
+                sol_item["r"] = ref(wm, :resistance, a)[r_id]
+            else
+                x_res, r_id = findmin(ref(wm, :resistance, a))
+                sol_item["r"] = ref(wm, :resistance, a)[r_id]
+            end
+        end
+    else
+        for (i, link) in data_dict
+            a = link["index"]
+            sol_item = sol_dict[i] = get(sol_dict, i, Dict{String, Any}())
+            x_res, r_id = findmin(ref(wm, :resistance, a))
+            sol_item["r"] = ref(wm, :resistance, a)[r_id]
+        end
+    end
+end
+
+"adds values based on JuMP variables"
+function add_setpoint!(
```

Comment:
Looking at this reminds me that we should setup some abstract solution building tools in InfrastructureModels, i.e. https://github.com/lanl-ansi/InfrastructureModels.jl/issues/1.  Feel free to remind me about next month.  :-)